[rocq] Support for installation of package metadata.

ejgallego opened the pull request:
We now install a `rocq-package` file for each Rocq theory. This way, when Rocq developments build with Dune are globally installed, we can find out additional metadata, such as the list of their dependencies.

Fixes #11012 fixes #11483

TODO:

- update changelog
- update documentation

Note: rebased on top of #12035 

Joint work with @Lysxia

Lysxia commented on src/dune_rules/coq/coq_lib.ml:
```diff
@@ -458,23 +458,23 @@ module DB = struct
     ;;
 
-    let create_from_stanza_impl (coq_db, db, dir, (s : Coq_stanza.Theory.t)) =
+    let create_from_coq_package_impl (coq_db, db, dir, (s : Coq_package.meta)) =
```
Our initial idea was to convert a `Coq_package` into a `Coq_stanza` but some fields don't seem easy to populate, so I am trying the other way around. However this memo table relies on physical equality of `Coq_stanza` so I'm not sure this still works as expected.

ejgallego replied:
I see! This is indeed not needed for the db, I'd suggest to make it an option for now, there are already some codepaths where we use get_exn for it.

ejgallego commented on src/dune_rules/coq/coq_lib.ml:
```diff
@@ -594,10 +602,9 @@ module DB = struct
-        let+ theory = create_from_coqpath ~boot_id cp in
-        Legacy theory
+        let db = assert false in
+        let dir = assert false in
```
dir and db are useful in the memo table as to distinguish stanzas that may appear (and be private) in different composed builds.

I think we can have a case where two identical stanzas live in two separate directories, if these are private. I'm not sure why `dir` is not enough here tho, I'd dare to say that stanzas are uniquely identified by the pair `(dir, stanza)`

We can ask Rudy once we get the patch more advanced, I guess for now we can pass `db = installed_db` and `dir` the directory we found the rocq-package file.

Lysxia commented:
I made some progress. Now it looks like I broke the finding of plugins:
```
File "test/blackbox-tests/test-cases/coq/compose-installed-plugin.t/run.t", line 1, characters 0-0:
...
+  File "$TESTCASE_ROOT/lib/coq/user-contrib/Bar/rocq-package", line 3, characters 9-22:
+  3 | (plugins global.plugin)
+               ^^^^^^^^^^^^^
+  Error: Library "global.plugin" not found.
+  -> required by theory Bar in
+     $TESTCASE_ROOT/lib/coq/user-contrib/Bar/rocq-package:2
+  -> required by _build/default/.Foo.theory.d
+  -> required by alias all
```

ejgallego commented:
> I made some progress. Now it looks like I broke the finding of plugins:
> 
> ```
> File "test/blackbox-tests/test-cases/coq/compose-installed-plugin.t/run.t", line 1, characters 0-0:
> ...
> +  File "$TESTCASE_ROOT/lib/coq/user-contrib/Bar/rocq-package", line 3, characters 9-22:
> +  3 | (plugins global.plugin)
> +               ^^^^^^^^^^^^^
> +  Error: Library "global.plugin" not found.
> +  -> required by theory Bar in
> +     $TESTCASE_ROOT/lib/coq/user-contrib/Bar/rocq-package:2
> +  -> required by _build/default/.Foo.theory.d
> +  -> required by alias all
> ```

I think this test-case is for the deprecated method of installing plugins in `user-contrib`, so IMHO it is fine if it broken.

You could fix it by scanning the cmxs files, but given that Rocq 9 doesn't support this anymore, IMHO not worth it.

ejgallego committed:
[rocq] Support for installation of package metadata.

We now install a `rocq-package` file for each Rocq theory. This way,
when Rocq developments build with Dune are globally installed, we can
find out additional metadata, such as the list of their dependencies.

Fixes #11012 fixes #11483

TODO:

- update changelog
- update documentation

Co-authored-by: Li-yao Xia <lysxia@gmail.com>
Signed-off-by: Emilio Jesus Gallego Arias <e+git@x80.org>

Lysxia commented:
@rgrinberg Why was this closed? I thought this is supposed to fix #11483

rgrinberg commented:
Let's re-open this once you start working on it. There's absolutely no pressure from me, and I am neither for nor against this feature. But there's no point using a Active PR's as a personal TODO list.
